network/discovery_test.go

```diff
@@ -98,3 +98,177 @@ func TestDiscovery_PeerAdded(t *testing.T) {
+		cancel()
+	})
+	if _, err := WaitUntilRoutingTableToBeFilled(ctx, servers[0], 1); err != nil {
+		t.Fatalf("server 0 should add a peer to routing table but didn't, peer=%s", servers[1].host.ID())
+	}
+	if _, err := WaitUntilRoutingTableToBeFilled(ctx, servers[1], 1); err != nil {
+		t.Fatalf("server 1 should add a peer to routing table but didn't, peer=%s", servers[0].host.ID())
+	}
+
+	assert.Contains(t, servers[0].discovery.routingTable.ListPeers(), servers[1].AddrInfo().ID)
+	assert.Contains(t, servers[1].discovery.routingTable.ListPeers(), servers[0].AddrInfo().ID)
+}
+
+func TestRoutingTable_Disconnected(t *testing.T) {
+	defaultConfig := &CreateServerParams{
+		ConfigCallback: func(c *Config) {
+			c.MaxPeers = 1
+		},
+	}
+	paramsMap := map[int]*CreateServerParams{
+		0: defaultConfig,
+		1: defaultConfig,
+	}
+
+	servers, createErr := createServers(3, paramsMap)
+	if createErr != nil {
+		t.Fatalf("Unable to create servers, %v", createErr)
+	}
+	t.Cleanup(func() {
+		closeTestServers(t, servers)
+	})
+
+	// connect to peer and make sure peer is in routing table
+	if joinErr := JoinAndWait(servers[0], servers[1], DefaultBufferTimeout, DefaultJoinTimeout); joinErr != nil {
+		t.Fatalf("Unable to join peers, %v", joinErr)
+	}
+
+	// make sure each routing table has peer
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	t.Cleanup(func() {
+		cancel()
+	})
+	if _, err := WaitUntilRoutingTableToBeFilled(ctx, servers[0], 1); err != nil {
+		t.Fatalf("server 0 should add a peer to routing table but didn't, peer=%s", servers[1].host.ID())
+	}
+	if _, err := WaitUntilRoutingTableToBeFilled(ctx, servers[1], 1); err != nil {
+		t.Fatalf("server 1 should add a peer to routing table but didn't, peer=%s", servers[0].host.ID())
+	}
+
+	// disconnect
+	if leaveErr := LeaveAndWait(servers[1], servers[0], DefaultLeaveTimeout); leaveErr != nil {
+		t.Fatalf("Unable to disconnect from server[0], err=%v", leaveErr)
+	}
+
+	// make sure each routing table remove a peer
+	ctx2, cancel2 := context.WithTimeout(context.Background(), 10*time.Second)
+	t.Cleanup(func() {
+		cancel2()
+	})
+	if _, err := WaitUntilRoutingTableToBeFilled(ctx2, servers[0], 0); err != nil {
+		t.Fatalf("server 0 should remove a peer from routing table but didn't, peer=%s", servers[1].host.ID())
+	}
+	if _, err := WaitUntilRoutingTableToBeFilled(ctx2, servers[1], 0); err != nil {
+		t.Fatalf("server 1 should remove a peer from routing table but didn't, peer=%s", servers[0].host.ID())
+	}
+}
+
+func TestRoutingTable_ConnectionFailure(t *testing.T) {
+	defaultConfig := &CreateServerParams{
+		ConfigCallback: func(c *Config) {
+			c.MaxPeers = 1
+		},
+	}
+	paramsMap := map[int]*CreateServerParams{
+		0: defaultConfig,
+		1: defaultConfig,
+	}
+
+	servers, createErr := createServers(3, paramsMap)
+	if createErr != nil {
+		t.Fatalf("Unable to create servers, %v", createErr)
+	}
+	t.Cleanup(func() {
+		// close only servers[0] because servers[0] has closed already
```

Comment:
I think you meant `[...] because server[1] has already closed`.
It seems like you're closing server[1] just below.

Reply:
Good catch! Thank you, I fixed at 20ae08d9f5c30f9a4e0fea7da8db05b7cc40bfb6